### PR TITLE
[Direct API] Update Direct API Access Config

### PR DESCRIPTION
**Pull request by billfienberg**
## WHY are these changes introduced?

These changes are being introduced to make the Direct API toml config more explicit.  

TLDR: 

```toml
[access.admin]
# Valid values include "online", and "offline"
# "online" is the default. 
direct_api_mode = "online"
# false is the default. 
embedded_app_direct_api_access = false
```

## WHAT is this pull request doing?

- Adds `access` to `accessConfiguration`
   - Adds `admin` inside `access`
   - Adds `directApiMode` inside `admin`
   - Adds `embedded_app_direct_api_access` inside `admin`
- Replaces the old config (`[access.direct_api_offline_access]`) with the new config (`[access.admin.direct_api_mode]`). 
- Adds `admin`, `directApiMode`, embeddedAppDirectApiAccess` to `AppAccessSchema`
- Passes `directApiMode` to the `dev` service
- Fixes some typos
- Adds/updates tests

## References

- Docs
   - [Direct API access](https://shopify.dev/docs/api/admin-extensions#direct-api-access)
   - [Access Modes](https://shopify.dev/docs/apps/auth/oauth/access-modes)
   - Action and Block Extension [Wiki](https://github.com/Shopify/app-ui/wiki/Action-and-Block-Extension)
   - https://vault.shopify.io/page/App-module-development-in-Spin~ToYp.md
   - [Google Doc: App config toml](https://docs.google.com/document/d/1EwN1mceHGdvepj8AxrgvchIlenzkiJs3tznfr_O1VSQ/edit?usp=sharing)
- RFCs
   - [Option 6](https://github.com/Shopify/ui-api-design/discussions/191#discussioncomment-8112987) from the CLI Config for API Access/Access Mode RFC
- PRs
   - https://github.com/Shopify/cli/pull/2802
   - https://github.com/Shopify/cli/pull/3258
   - https://github.com/Shopify/shopify/pull/476372
   - https://github.com/Shopify/web/pull/115252
   - https://github.com/Shopify/web/pull/115621

## Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

## Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/models/app/loader.ts [hunk 1]
```diff
@@ -429,8 +429,8 @@ class AppLoader {
           this.abortOrReport.bind(this),
         )
 
-        const {path, ...specConfigurationWithouPath} = specConfiguration
-        if (Object.keys(specConfigurationWithouPath).length === 0) return
+        const {path, ...specConfigurationWithoutPath} = specConfiguration
```
Comment (billfienberg) [inline packages/app/src/cli/models/app/loader.ts:432]: Fixed typo (`specConfigurationWithouPath` => `specConfigurationWithoutPath`)

**Comment by billfienberg** [comment]
I temporarily added the following npm scripts in my local `cli` repo so I didn't have to remember/type out all the vars/args/flags/etc. 

```json
"temp:deploy": "SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 cli-spin pnpm shopify app deploy --reset --path ~/src/github.com/Shopify/2024-01-22-project",
"temp:link": "SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 cli-spin pnpm shopify app config link --reset --path ~/src/github.com/Shopify/2024-01-22-project",
"temp:dev": "SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 cli-spin pnpm shopify app dev --reset --path ~/src/github.com/Shopify/2024-01-22-project",
"temp:generate-extension": "cli-spin pnpm shopify app generate extension --reset --path ~/src/github.com/Shopify/2024-01-22-project"
```

File: packages/app/src/cli/models/extensions/specifications/app_config_app_access.ts [hunk 2]
```diff
@@ -5,7 +5,9 @@ import {zod} from '@shopify/cli-kit/node/schema'
 const AppAccessSchema = zod.object({
   access: zod
     .object({
-      direct_api_offline_access: zod.boolean().optional(),
+      admin: zod
+        .object({direct_api_mode: zod.union([zod.literal('disabled'), zod.literal('online'), zod.literal('offline')])})
+        .optional(),
     })
     .optional(),
```
Comment (billfienberg) [inline packages/app/src/cli/models/extensions/specifications/app_config_app_access.ts:12]: We went through a Config API Alignment process, and originally settled on an API that looked like checkout's  `api_access` [config](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/configuration#how-it-works) (Checkout: `[extensions.capabilities.api_access]`, Option 5 from RFC: `[capabilities.api_access.admin]`). But after some concerns about `[capabilities]` being too vague at the app level and `[access]` being more explicit, I decided to move forward with `[access]`. 

More info: https://github.com/Shopify/ui-api-design/discussions/191#discussioncomment-8112987

Comment (pt2pham) [reply packages/app/src/cli/models/extensions/specifications/app_config_app_access.ts:12]: Happy with Option 6. Overall very similar to the agreed upon Option 5, but is more accurate and specific at the app root level. Overall I don't think it's too much of a departure from the Option 5 proposal for this to be a big difference. It still follows the root of `access` -> `admin` and is now explicit about the `direct_api` feature.

File: packages/app/src/cli/models/app/app.ts [hunk 3]
```diff
@@ -314,11 +314,23 @@ export class App implements AppInterface {
...
+      'access.admin.embedded_app_direct_api_access',
+    )
     return {
       ...(scopes || useLegacyInstallFlow
         ? {access_scopes: {scopes, use_legacy_install_flow: useLegacyInstallFlow}}
         : {}),
       ...(redirectUrls ? {auth: {redirect_urls: redirectUrls}} : {}),
+      ...(directApiMode
+        ? {
+            access: {
```
Comment (billfienberg) [inline packages/app/src/cli/models/app/app.ts:329]: TODO: Test direct_api_mode with falsy embedded_app_direct_api_access, and vice versa.

**Comment by billfienberg** [comment]
## Test Cases

### From Local App to Spin DB

```toml
[access.admin]
direct_api_mode = "online"
embedded_app_direct_api_access = true
```

```toml
[access.admin]
direct_api_mode = "online"
embedded_app_direct_api_access = false
```

```toml
[access.admin]
direct_api_mode = "offline"
embedded_app_direct_api_access = true
```

```toml
[access.admin]
direct_api_mode = "offline"
embedded_app_direct_api_access = false
```

### From Spin DB to Local App

**Review by alfonso-noriega — APPROVED** [review]
🎩 LGTM
